File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.Stack;
import java.io.FileInputStream;


class Solution
{
	public static void main(String args[]) throws Exception
	{
		Scanner sc = new Scanner(System.in);
		int T;
		T=10;

		for(int test_case = 1; test_case <= T; test_case++)
		{
			int N = sc.nextInt();
			sc.nextLine();
			String str = sc.nextLine();
			Stack<Character> stack = new Stack<>();
			List<Character> left = new ArrayList<>();
			left.add('(');
			left.add('{');
			left.add('[');
			left.add('<');
			List<Character> right = new ArrayList<>();
			right.add(')');
			right.add('}');
			right.add(']');
			right.add('>');
//			array_left = Arrays.asList(left);
			for (int i = 0; i<str.length();i++) {
				char text = str.charAt(i);
				if (left.contains(text)) {
					stack.push(text);
				}
				if (right.contains(text)) {
					if (right.indexOf(text) == left.indexOf(stack.peek())) {
						stack.pop();
					}
					else {
						break;
					}
					
				}
			}
			int answer = 0;
			if (stack.size() == 0) {
				answer = 1;
			}
			System.out.println("#" + test_case + " " + answer);
		}
        sc.close();
	}
}

File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.py
T = 10
for test_case in range(1, T+1):
    N = int(input())
    string = input()
    stack = []
    # 여는 괄호
    left = ['(', '{', '[', '<']
    # 닫는 괄호
    right = [')', '}', ']', '>']
    for i in range(len(string)):
        # 여는 괄호면 그냥 추가
        if string[i] in left:
            stack.append(string[i])
        if string[i] in right:
            # 가장 상위의 괄호 값과 쌍이라면
            if right.index(string[i]) == left.index(stack[-1]):
                # 상위의 원소 제거하기
                stack.pop()
            else:
                break

    answer = 0
    if len(stack) == 0:
        answer = 1
    print(f"#{test_case} {answer}")